src/menu/charprofile.c

```diff
@@ -0,0 +1,409 @@
+	},
+	[PROFILE_LOCKED] = {
+		.name = "locked",
+		.fullname = "Locked",
+		.title = "...",
+		.description = "You have not unlocked this character yet!",
+		.unlock = PBGM_GAMEOVER,
+		.background = "stage1/cirnobg",
+		.faces = FACES("locked"),
+	}
+};
+
+
+static void update_char_draw_order(MenuData *m) {
+	CharProfileContext *ctx = m->context;
+
+	for(int i = 0; i < NUM_PROFILES; ++i) {
+		if(ctx->char_draw_order[i] == m->cursor) {
+			while(i) {
+				ctx->char_draw_order[i] = ctx->char_draw_order[i - 1];
+				ctx->char_draw_order[i - 1] = m->cursor;
+				--i;
+			}
+
+			break;
+		}
+	}
+}
+
+static int check_unlocked_profile(int i) {
+	int selected = PROFILE_LOCKED;
+	if(!profiles[i].unlock) {
+		// for protagonists
+		selected = i;
+	} else if(profiles[i].unlock != PBGM_GAMEOVER) {
+		if(progress_is_bgm_id_unlocked(profiles[i].unlock)) selected = i;
+	}
+	return selected;
+}
+
+static void charprofile_logic(MenuData *m) {
+	dynarray_foreach(&m->entries, int i, MenuEntry *e, {
+		e->drawdata += 0.05 * ((m->cursor != i) - e->drawdata);
+	});
+	MenuEntry *cursor_entry = dynarray_get_ptr(&m->entries, m->cursor);
+	Font *font = res_font("small");
+	char buf[512] = { 0 };
+	int j = check_unlocked_profile(m->cursor);
+	text_wrap(font, profiles[j].description, DESCRIPTION_WIDTH, buf, sizeof(buf));
+	double height = text_height(font, buf, 0) + font_get_lineskip(font) * 2;
+
+	fapproach_asymptotic_p(&m->drawdata[0], 1, 0.1, 1e-5);
+	fapproach_asymptotic_p(&m->drawdata[1], 1 - cursor_entry->drawdata, 0.1, 1e-5);
+	fapproach_asymptotic_p(&m->drawdata[2], height, 0.1, 1e-5);
+}
+
+static void charprofile_draw(MenuData *m) {
+	assert(m->cursor < NUM_PROFILES);
+	r_state_push();
+
+	CharProfileContext *ctx = m->context;
+
+	draw_main_menu_bg(m, SCREEN_W/4+100, 0, 0.1 * (0.5 + 0.5 * m->drawdata[1]), "menu/mainmenubg", profiles[m->cursor].background);
+	draw_menu_title(m, "Character Profiles");
+	draw_menu_list(m, 100, 100, NULL, SCREEN_H);
+
+	r_mat_mv_push();
+	r_mat_mv_translate(SCREEN_W/4, SCREEN_H/3, 0);
+
+	float f = m->drawdata[0];
+	float descbg_ofs = 100 + 30 * f - 20 * f - font_get_lineskip(res_font("small")) * 0.7;
+
+	r_color4(0, 0, 0, 0.5);
+	r_shader_standard_notex();
+	r_mat_mv_translate(-150, descbg_ofs + m->drawdata[2] * 0.5, 0);
+	r_mat_mv_scale(650, m->drawdata[2], 1);
+	r_draw_quad();
+	r_shader_standard();
+	r_mat_mv_pop();
+
+	CharProfiles i = ctx->char_draw_order[m->cursor];
+
+	MenuEntry *e = dynarray_get_ptr(&m->entries, m->cursor);
+
+	float o = 1 - e->drawdata*2;
+	float pbrightness = 0.6 + 0.4 * o;
+
+	float pofs = fmax(0.0f, e->drawdata * 1.5f - 0.5f);
+	pofs = glm_ease_back_in(pofs);
+
+	if(i != m->selected) {
+		pofs = lerp(pofs, 1, menu_fade(m));
+	}
+
+	int selected = check_unlocked_profile(m->cursor);
+
+	Color *color = RGBA(pbrightness, pbrightness, pbrightness, 1);
+	// if not unlocked, darken the sprite so it's barely visible
+	if(selected == PROFILE_LOCKED) color = RGBA(0.0, 0.0, 0.0, 0.9);
```

Comment:
Code style nitpick - this should be expanded into the standard block form. One-line `if`s may sometimes be acceptable when you have multiple short and similar `if` statements, but it's usually frowned upon.